HIP/hip-336.md

```diff
@@ -0,0 +1,301 @@
+		bool receiverSigRequired = 11;
+		Timestamp expirationTime = 12;
+		Duration autoRenewPeriod = 13;
+		repeated LiveHash liveHashes = 14;
+		repeated TokenRelationship tokenRelationships = 15;
+		string memo = 16;
+		int64 ownedNfts = 17;
+		int32 max_automatic_token_associations = 18;
+
+		// new fields
+		int32 totalAllowances = 19;
+		repeated CryptoAllowance allowances = 20;
+		repeated TokenAllowance tokenAllowances = 21;
+	}
+
+	// existing fields
+	ResponseHeader header = 1;
+	AccountInfo accountInfo = 2;
+}
+```
+
+The `TokenAllowance` message will contain the allowance information for a single fungible or non-fungible token spendable by the `spender`.
+
+If the token type is non-fungible, the `approvedForAll` field may be set to `true` if the spender has been granted access to all instances of the NFT owned by the owner. Fungible tokens will not `approvedForAll` as this field is only relevant for NFTs.
+
+The `allowances` and `approvedForAll` fields in the `TokenAllowance` message are specified using wrapper types so callers can determine whether the field has a value or not, as opposed to inferring from the type default value (for example, 0 or false).
+
+The `decimals` field, which specifies the number of decimal places the token is divisible by, is returned as a courtesy to the caller so another call to `TokenGetInfo` is not required. This field will have value `0` for NFTs.
+
+Only allowances with a non-zero value will be returned.
+
+The `spender`'s allowance will change as tokens are spent or additional allowances are approved by the token owner.
+
+The current number of allowances approved by the owner is specified in `totalAllowances`. This value is calculated as the size of `allowances` plus the size of `tokenAllowances`.
+
+### Add CryptoApproveAllowance API
+
+A new Hedera API will be added under the Crypto Service called `CryptoApproveAllowance`. This function is called by the token owner to create a series of allowances for a spender account.
+
+```protobuf
+message CryptoApproveAllowanceTransactionBody {
+	message CryptoApproval {
+		AccountID spender = 1;
+		int64 amount = 2;
+	}
+
+	message TokenApproval {
+		TokenID tokenId = 1;
+		AccountID spender = 2;
+		oneof allowance {
+			int64 amount = 1;
+			bool approvedForAll = 2;
+		}
+	}
+
+	repeated CryptoApproval cryptoApproval = 1;
+	repeated TokenApproval tokenApproval = 2;
+}
+```
+
+This function contains a list of hbar and/or token approval messages. Each of these messages will set an allowance of `amount` tokens or hbar that can be transferred from the owner's account to another account by the `spender` account. Each of the hbar/token approvals in the list do not need to refer to the same spender. The owner can establish different allowances for various spenders in a single transaction.
+
+The total number of approvals contained within the transaction body cannot exceed 100. This number includes both hbar and token approvals.
```

Comment:
Is this number 100 the same as for crypto transfers? I heard today crypto transfers was limited to 20 transfer items. Should this number align with that one? Same question for line 150

Reply:
I aligned the max list size to the max number of allowances per acct (100). But if crypto transfers has a max list size of 20, I'll make this the same. The max allowances per acct will still be 100.